File: src/main/java/com/celeste/tasks/model/list/RunnableList.java
package com.celeste.tasks.model.list;

import com.google.common.collect.ForwardingList;
import java.util.Collection;
import java.util.List;
import java.util.function.Supplier;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;

@RequiredArgsConstructor
public class RunnableList<T> extends ForwardingList<T> {

  private final List<T> list;
  private final Runnable runnable;

  public RunnableList(final Supplier<List<T>> supplier, final Runnable runnable) {
    this(supplier.get(), runnable);
  }

  public boolean addDefault(final T result) {
    return super.add(result);
  }

  public boolean addAllDefault(final Collection<? extends T> collection) {
    return super.addAll(collection);
  }

  @Override
  public boolean add(@NotNull final T result) {
    final boolean status = super.add(result);
    final Thread thread = new Thread(runnable);
    thread.setDaemon(true);
    thread.start();

    return status;
  }

  @Override
  public boolean addAll(@NotNull final Collection<? extends T> collection) {
    final boolean status = super.addAll(collection);
    final Thread thread = new Thread(runnable);
    thread.setDaemon(true);
    thread.start();

    return status;
  }

  @Override
  protected List<T> delegate() {
    return list;
  }

}

File: src/main/kotlin/com/cefoler/tasks/model/list/RunnableList.kt
package com.cefoler.tasks.model.list

import com.google.common.collect.ForwardingList
import lombok.RequiredArgsConstructor
import java.util.function.Supplier

@RequiredArgsConstructor
class RunnableList<T>(
  private val list: List<T>,
  private val runnable: Runnable,
) : ForwardingList<T>() {

  constructor(supplier: Supplier<List<T>>, runnable: Runnable) : this(supplier.get(), runnable)

  fun addDefault(result: T): Boolean {
    return super.add(result)
  }

  fun addAllDefault(collection: Collection<T>): Boolean {
    return super.addAll(collection)
  }

  override fun add(result: T): Boolean {
    val status = super.add(result)
    val thread = Thread(runnable)

    thread.isDaemon = true
    thread.start()

    return status
  }

  override fun addAll(collection: Collection<T>): Boolean {
    val status = super.addAll(collection)
    val thread = Thread(runnable)

    thread.isDaemon = true
    thread.start()

    return status
  }

  override fun delegate(): List<T> {
    return list
  }

}